clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,464 @@
+ *
+ * @param <K> The key type, which is also the granularity of the request routing (e.g.
+ *            this could be `TopicPartition` in the case of requests intended for a partition
+ *            leader or the `GroupId` in the case of consumer group requests intended for
+ *            the group coordinator)
+ * @param <V> The fulfillment type for each key (e.g. this could be consumer group state
+ *            when the key type is a consumer `GroupId`)
+ */
+public class AdminApiDriver<K, V> {
+    private final Logger log;
+    private final long retryBackoffMs;
+    private final long deadlineMs;
+    private final AdminApiHandler<K, V> handler;
+    private final Optional<StaticKeyMapping<K>> staticMapping;
+    private final Optional<DynamicKeyMapping<K>> dynamicMapping;
+    private final Map<K, KafkaFutureImpl<V>> futures;
+
+    private final BiMultimap<RequestScope, K> lookupMap = new BiMultimap<>();
+    private final BiMultimap<BrokerScope, K> fulfillmentMap = new BiMultimap<>();
+    private final Map<RequestScope, RequestState> requestStates = new HashMap<>();
+
+
+    public AdminApiDriver(
+        AdminApiHandler<K, V> handler,
+        long deadlineMs,
+        long retryBackoffMs,
+        LogContext logContext
+    ) {
+        this.handler = handler;
+        this.deadlineMs = deadlineMs;
+        this.retryBackoffMs = retryBackoffMs;
+        this.log = logContext.logger(AdminApiDriver.class);
+        this.futures = new HashMap<>();
+
+        KeyMappings<K> result = handler.initializeKeys();
+        this.dynamicMapping = result.dynamicMapping;
+        this.staticMapping = result.staticMapping;
+
+        initializeKeys();
+    }
+
+    private void initializeKeys() {
+        staticMapping.ifPresent(mapping -> {
+            for (Map.Entry<K, Integer> staticMapping : mapping.keys.entrySet()) {
+                K key = staticMapping.getKey();
+                Integer brokerId = staticMapping.getValue();
+                futures.put(key, new KafkaFutureImpl<>());
+                map(key, brokerId);
+            }
+        });
+
+        dynamicMapping.ifPresent(mapping -> {
+            for (K key : mapping.keys) {
+                futures.put(key, new KafkaFutureImpl<>());
+                lookupMap.put(mapping.lookupStrategy.lookupScope(key), key);
+            }
+        });
+    }
+
+    /**
+     * Associate a key with a brokerId. This is called after a response in the Lookup
+     * stage reveals the mapping (e.g. when the `FindCoordinator` tells us the the
+     * group coordinator for a specific consumer group).
+     */
+    private void map(K key, Integer brokerId) {
+        lookupMap.remove(key);
+        fulfillmentMap.put(new BrokerScope(brokerId), key);
+
+        // To allow for derived keys, we create futures dynamically if they
+        // do not already exist in the future map
+        futures.computeIfAbsent(key, k -> new KafkaFutureImpl<>());
```

Comment:
As `futures` is already returned to callers (`DescribeProducersResult`), is it thread-safe if we add more futures dynamically?

Reply:
Good question. The notion of a derived key came from trying to do the `ListTransactions` handler. This request is intended to be sent to all brokers in the cluster, so I used the brokerId as a derived key: https://github.com/apache/kafka/pull/9268/files#diff-046707392764feb1bf0126ee719554716865a0497741b3d721e069cf10a0d323R40. First we send `Metadata` to get the list of brokers, then we send one `ListTransactions` request to each broker. It was safe in this case because the result type represented a single future which returned a map of futures: https://github.com/apache/kafka/pull/9268/files#diff-8a62abec7d0e85d1aa7faf7852b92cf6570666477cef5644b307166eeb69a361R39. I will think a little bit about how to represent this more clearly in the API. One possibility is to take it out for now and save it for when I submit the `ListTransactions` PR. I might just do that since I haven't yet added any testing for dynamic keys.